src/data/FileAdapter.js

```diff
@@ -108,6 +108,37 @@ export default class {
 
     return game;
   }
+
+  async cancelGame(game) {
+
+    return this._lock(`game_${game.id}`, 'write', async () => {
+      let gameRefreshed = await this._readFile(`game_${game.id}`, null);
+      if (gameRefreshed.state.started) {
```

Comment:
If, somehow, the client manages to submit 2 near-parallel cancel game requests for the same game, then I expect one of them will call this method first, and the other one will wait for a lock.  The 2nd call will then attempt to read the file.  But the file doesn't exist.  It won't throw an error because you passed an initial value of NULL.  This is good.  You don't want to throw an exception if the file doesn't exist.  But it will, because you are assuming "gameRefreshed" has a value.  Instead, I think the right thing to do would be to return "true" if gameRefreshed is null.  This makes the cancel/delete operation idempotent.

Reply:
Added return true if gameRefreshed is null, also added check to GameService to make sure it doesn't try to delete gamePara which has already might have already been deleted by the first request.